pup-0003.md

```diff
@@ -0,0 +1,154 @@
+---
+
+## Summary
+
+This PUP is to redefine our git workflow to use cherry-picking instead of
+merging commits forward from x.y-dev branches. Using this model, Pull Requests
+(PRs) would be made against master and then bug fixes would be cherry-picked to
+the older x.y-dev branch as needed. Hotfixes would be cherry-picked from
+x.y-release branch into the x.y-dev branch and master.
+
+## Motivation
+
+Currently in Pulp, we merge features to master. However, bug fixes are merged
+to x.y-dev branches which then are merged to master (a process called "merging
+forward"). More info about merging forward can be found in our documentation:
+
+http://docs.pulpproject.org/en/2.12/dev-guide/contributing/merging.html#merging-your-changes
+
+Also, a good illustration of this process is outlined in the following webpage:
+
+http://nvie.com/posts/a-successful-git-branching-model/
+
+As a result, our git history is complex with multiple branches feeding into one
+another and numerous merge commits. Using cherry-picks would simplify our git
+history by keeping it more linear and eliminating merge commits.
+
+Opening a PR requires knowledge of our current process. This can create a
+barrier to entry for community members. With the cherry-pick model, all PRs are
+opened against master and the onus to update older x.y-dev branches is on the
+person merging (who usually has better knowledge of which branches a commit
+needs to be applied to).
+
+We've also seen some serious mistakes where branches are accidentally merged
+into other branches. These situations have occurred multiple times and each time
+required multiple people working several hours to resolve the problem. After
+being resolved, the git history is convoluted with huge negative commits which
+later caused confusion at release time. Commits could be accidentally
+cherry-picked, but this would be less disastrous since it's only a single
+commit.
+
+Changes that we want to only be made to older x.y-dev branches (but not to
+newer ones) still need to be merged forward. We could elminate the need for
+this with cherry-picking. It's worth noting though that this rarely happens.
+
+## Detailed Design
+
+This proposal would keep the current branching scheme[1] involving master,
+x.y-dev branches, and x.y-release branches. The workflow by which changes are
+moved between the branches would however change. The cherry picking process will
+apply to Pulp2, Pulp3, and future Pulp versions.
+
+The process works a bit differently depending on whether the change is a new
+feature, bug fix, or hotfix.
+
+#### New Feature
+
+1. All feature PRs are opened and merged against master. No cherry picking will
+   occur since new features only land on master.
+
+#### Bug Fix
+
+1. All bug fix PRs are opened against master
+2. After the PR is ready to be merged, the person merging the commit would
+   first merge the commit to master. Then, this person cherry-picks this change
+   from master to the latest x.y-dev branch.
```

Comment:
If we make it 100% via PR (even if the PR needs no review) then we can disable pushing on branches entirely.